# Define UI for MetaAnalyst application
ui <- dashboardPage(
  #define the skin
  skin = "green",
  
  dashboardHeader(title = "TransciptMetaAnalyst"),
  #close the header
  
  ###########################
  #####-----SIDEBAR-----#####
  ###########################
  dashboardSidebar(sidebarMenu(
    id = "main_tabs",
    #create a home tab to explain the app
    menuItem("Home",
             tabName = "home",
             icon = icon("home"),
             selected = T), #close home tab
    
    #create a RNA-Seq tab
    convertMenuItem(menuItem("RNA-Seq Meta Analysis",
                             tabName = "rnaseq",
                             icon = icon("list-alt", lib = "glyphicon"),
                             # upload button for counts
                             fileInput(inputId = "rnaseq_counts",
                                       label = "Upload the raw counts file (tab-delimited)",
                                       accept = c("text/csv", "text/comma-separated-values,text/plain", "txt")),
                             #upload button for experimental design
                             fileInput(inputId = "rnaseq_design",
                                       label = "Upload the experimental design file (tab-delimited)",
                                       accept = c("text/csv", "text/comma-separated-values,text/plain", ".txt")),
                             #example files
                             downloadLink("example_counts", label = "Download example counts file \n"),
                             p(),
                             downloadLink("example_design", label = "Download example design file"),
                             #add button to start analysis
                             actionButton("rnaseq_analyze", "Start Analysis!", icon = icon("laptop"))
                             ),
                    tabName = "rnaseq"), #close rna-seq tab
    
    #create a tab for demo analysis
    menuItem("Run Demo Analysis",
             tabName = "demo",
             icon = icon("person-running")),
    
    #create a tabe for user manual
    menuItem("User Manual",
             tabName = "manual",
             icon = icon("book"))
  )),  #close the sidebar
  
  ########################
  #####-----BODY-----#####
  ########################
  dashboardBody(
    useShinyjs(),
    
    tabItems(
      ##### HOME TAB #####
      ####################
      # add box explaining the application
      tabItem(tabName = "home", class = "active",
        fluidRow(
          box(title = "Quick overview",
            h3(strong("TransciptMetaAnalyst"), ": A graphical user interface to meta-analyze RNA-Seq transcriptomics data"),
            p("This app is designed to help biologists with less programming skills to perform meta-analysis of several transcriptomic (either RNA-Seq-generated datasets in order to get robust results and new insights into the data that was generated by their own studies or the data available online."),
            h4("If you would like to check the app, please click on", strong("Run Demo Analaysis"), "tab..."),
            hr(),
            p("As shown on the side panel, the app consists of several tabs as follows:"),
            tags$ul(
              tags$li(strong("Home:"), "contains the preliminary information about the app and information about how to use it."),
              tags$li(strong("RNA-Seq Meta Analysis:"), "is designed to perform meta-analysis for a group of RNA-Seq-generated transcriptomic datasets."),
              tags$li(strong("Run Demo Analysis:"), "could be used to perform a sample RNA-Seq meta analysis on a dataset provided with the app. When the user clicks this tab, the analysis will start automatically and the results will appear on the screen after 5-10 seconds. The provided figures and tables are examples of what the application will provide under RNA-Seq Meta Analysis tab using the user's uploaded files"),
              tags$li(strong("User Manual:"), "contains a detailed user manual to guide the user during the app usage and explain all the files required to perform the analysis. It also contains information on how and/or where to get help, and how to cite the application.")
            ),
            solidHeader = T,
            width = 12,
            status = "primary")
        )
      ), #close home tab item
      
      ##### RNA-SEQ ANALYSIS TAB #####
      ################################
      tabItem(tabName = "rnaseq",
        div(id = "rnaseq_steps",
            fluidPage(
              box(title = "RNA-Seq Quick Start",
                  p("To start the RNA-Seq meta analysis please perform the following steps:",
                    tags$ul(
                      tags$li("After cleaning the raw RNA-Seq read files (either downloaded from public datbases of generated via your own experiments) of each sample, align them individually to the studied organism's reference genome."),
                      tags$li("Count the mapped reads to the reference genomic features of interest and combine all the counts in one file. Keeping genomic features as rows and the samples that will be included in the analysis as columns. To produce the final raw (unnormalized) counts comma-separated file.", strong("Note that the counts file should have a 'geneid' column.")),
                      tags$li("Upload the ", strong("counts"),"file produced by any feature counting tool, e.g., featureCounts or htseq-count, as exaplined above to the raw counts placeholder.", strong(" Please refer to example counts file via the provided download link.")),
                      tags$li("Upload the experimental design file.", strong(" Please refer to the example file via the procided download link."), "This file should contain at least 3 columns, namely 'sample' refers to the sample names used as columns in the counts file, 'study' refers to the study to which the sample belong, 'condition' referes to the treatment or control conditions (only two conditions allowed for comparison)."),
                      tags$li("Hit",strong('Start Analysis!'), "button and wait for the results to appear.")
                    )),
                  solidHeader = T, width = 12, status = "info")
            )), #close steps box
        #create general results and information section
        shinyjs::hidden(div(id = "general_results_rnaseq",
                            fluidRow(
                              #number of studies
                              infoBoxOutput("num_studies_rnaseq", width = 3),
                              #download of individual DEGs
                              box(title = "Download DEGs of individual studies", solidHeader = T, width = 4, height = 90, status = "primary",
                                  column(7, selectInput("download_individual_rnaseq", label = NULL,
                                                        choices = "", multiple = F)),
                                  column(4, downloadButton("download_individual_button_rnaseq", label = "Download DEGs"))),
                              #number of identified DEGs box
                              infoBoxOutput("num_degs_rnaseq", width = 5)
                            ))), #close general results section
        #create plots section
        shinyjs::hidden(div(id= "mainplots",
                            fluidRow(
                              #create a tab box for the main plots
                              tabBox(title = "Main Plots",
                                     tabPanel(title = "p-Values Dist", plotOutput("p_dist_rnaseq", height = "480px"),
                                              downloadButton("download_pdist_rnaseq", label = "Download plot")),
                                     tabPanel(title = "PCA", plotOutput("pca_rnaseq", height = "480px"),
                                              downloadButton("download_pca_rnaseq", label = "Download plot")),
                                     tabPanel(title = "Heatmap", d3heatmapOutput("heatmap_rnaseq", height = "480px"),
                                              downloadButton("download_heatmap_rnaseq", label = "Download plot"),
                                              downloadButton("download_heatmap_data_rnaseq", label = "Download data")),
                                     tabPanel(title = "Venn diagram", plotOutput("venn_rnaseq", height = "480px"),
                                              downloadButton("download_venn_rnaseq", label = "Download plot")),
                                     tabPanel(title = "UpSet Plot", plotOutput("upset_rnaseq", height = "480px"),
                                              downloadButton("download_upset_rnaseq", label = "Download plot")),
                                     tabPanel(title = "Volcano Plot", plotOutput("volcano_rnaseq", height = "480px"),
                                              downloadButton("download_volcano_rnaseq", label = "Download plot"),
                                              downloadButton("download_volcano_data_rnaseq", label = "Download data"))), #close the main plots box
                              #create a box for DEGs table
                              box(title = "Meta-analysis DEGs", status = "success", solidHeader = T,
                                  dataTableOutput("rnaseq_DEGs"),
                                  downloadButton("download_meta_degs_rnaseq", label = "Download table"))
                            ))),
        #create Go and counts section
        shinyjs::hidden(div(id= "GO_counts",
                            fluidRow(
                              #create a tab box for GO and KEGG analysis
                              tabBox(title = "Functional Annotation",
                                     tabPanel(title = "Gene Ontology",
                                              selectizeInput("organism_go_rnaseq", "Select organism", choices = "", multiple = F),
                                              actionButton("run_go_rnaseq", "Run GO enrichment", icon = icon("person-running")),
                                              plotOutput("go_plot_rnaseq", height = "450px") %>% withSpinner(type = 6),
                                              downloadButton("download_go_table_rnaseq", label = "Download GO table"),
                                              downloadButton("download_go_plot_rnaseq", label = "Download GO plot")
                                     ),
                                     tabPanel(title = "KEGG pathway",
                                              textInput("organism_kegg_rnaseq", 
                                                        label = p("Enter organism code found at", tags$a(
                                                          "KEGG database", target = "_blank", href = "https://www.genome.jp/kegg-bin/get_htext?br08601"))),
                                              actionButton("run_kegg_rnaseq", "Run KEGG enrichment", icon = icon("person-running")),
                                              plotOutput("kegg_plot_rnaseq", height = "450px") %>% withSpinner(type = 6),
                                              downloadButton("download_kegg_table_rnaseq", label = "Download KEGG table"),
                                              downloadButton("download_kegg_plot_rnaseq", label = "Download KEGG plot"))),
                              box(title = "Count plots", status = "primary", solidHeader = T,
                                  textOutput("counts_message_rnaseq"),
                                  plotOutput("counts_plots_rnaseq", height = "550px"),
                                  downloadButton("download_counts_plot_rnaseq", label = "Download counts plot"))
                            )))
      ), #close rnaseq tab item
      ##### DEMO ANALYSIS TAB #####
      #############################
      tabItem(tabName = "demo",
              #create general results and information section
              fluidRow(
                #number of studies
                infoBoxOutput("num_studies_demo", width = 3),
                #download of individual DEGs
                box(title = "Download DEGs of individual studies", solidHeader = T, width = 4, height = 90, status = "primary",
                    column(7, selectInput("download_individual_demo", label = NULL,
                                          choices = "", multiple = F)),
                    column(4, downloadButton("download_individual_button_demo", label = "Download DEGs"))),
                #number of identified DEGs box
                infoBoxOutput("num_degs_demo", width = 5)
              ), #close general results section
              #create plots section
              fluidRow(
                tabBox(title = "Main Plots",
                       #p-value distribution box
                       tabPanel(title = "p-Values Dist", plotOutput("p_dist_demo", height = "480px"),
                                downloadButton("download_pdist_demo", label = "Download plot")),
                       tabPanel(title = "PCA", plotOutput("pca_demo", height = "480px"),
                                downloadButton("download_pca_demo", label = "Download plot")),
                       tabPanel(title = "Heatmap", d3heatmapOutput("heatmap_demo", height = "480px"),
                                downloadButton("download_heatmap_demo", label = "Download plot"),
                                downloadButton("download_heatmap_data_demo", label = "Download data")),
                       tabPanel(title = "Venn diagram", plotOutput("venn_demo", height = "480px"),
                                downloadButton("download_venn_demo", label = "Download plot")),
                       tabPanel(title = "UpSet Plot", plotOutput("upset_demo", height = "480px"),
                                downloadButton("download_upset_demo", label = "Download plot")),
                       tabPanel(title = "Volcano Plot", plotOutput("volcano_demo", height = "480px"),
                                downloadButton("download_volcano_demo", label = "Download plot"),
                                downloadButton("download_volcano_data_demo", label = "Download data"))),
                box(title = "Meta-analysis DEGs", status = "success", solidHeader = T,
                    dataTableOutput("demo_DEGs"),
                    downloadButton("download_meta_degs_demo", label = "Download table")),
              ),
              #create DataTable of the identified DEGs
              fluidRow(
                #create a tab box for GO and KEGG analysis
                tabBox(title = "Functional Annotation",
                       tabPanel(title = "Gene Ontology",
                                actionButton("run_go_demo", "Run GO enrichment", icon = icon("person-running")),
                                plotOutput("go_plot_demo") %>% withSpinner(type = 6),
                                downloadButton("download_go_table_demo", label = "Download GO table"),
                                downloadButton("download_go_plot_demo", label = "Download GO plot")
                       ),
                       tabPanel(title = "KEGG pathway",
                                actionButton("run_kegg_demo", "Run KEGG enrichment", icon = icon("person-running")),
                                plotOutput("kegg_plot_demo") %>% withSpinner(type = 6),
                                downloadButton("download_kegg_table_demo", label = "Download KEGG table"),
                                downloadButton("download_kegg_plot_demo", label = "Download KEGG plot"))),
                box(title = "Count plots", status = "primary", solidHeader = T,
                    textOutput("counts_message_demo"),
                    plotOutput("counts_plots_demo"),
                    downloadButton("download_counts_plot_demo", label = "Download counts plot"))
              )
      ), #close demo tab item
      
      ##### USER GUIDE TAB #####
      ##########################
      tabItem(tabName = "manual",
              fluidPage(
                box(title = "Detailed User Guide",
                    solidHeader = T, width = 12, status = "danger",
                    h3("TransciptMetaAnalyst: Datailed User Manual"),
                    p("A detailed user manual could be downloaded via this link."), ##### ADD LINK OF PDF FILE
                    hr(),
                    h3("Contact us"),
                    span("For inquiries and/or feedback, please contact me via"),
                    a("email",
                      href = "mailto:e.abdelsalam@lmu.de"),
                    hr(),
                    h3("How to cite?"),
                    p("Please cite TransciptMetaAnalyst as follows:")
                    )
              )
      ) #close guide tab item
    ), #close all tab items
    ##################################################################################################################
    
    #add footer and version
    tags$footer(
      tags$p("\U00A9 Developed and supported by Eslam M. Abdel-Salam, Ludwig Maximilians University of Munich (LMU)"),
      align = "right"), # Dasbboardbody close 
    shiny.info::version(position = "bottom right")
    
  ) #close the body
) #close the whole page
